buildSrc/src/main/kotlin/codegen/math/GenerateMath.kt

```diff
@@ -0,0 +1,249 @@
+	@OutputDirectory
+	val outputDir = project.objects.directoryProperty()
+
+	@get:OutputDirectory
+	val commonDir = outputDir.file("common")
+
+
+	private val packageName = "com.kgl.math"
+
+	private val typeList = listOf(
+		BYTE to "0",
+		SHORT to "0",
+		INT to "0",
+		LONG to "0L",
+		U_BYTE to "0U",
+		U_SHORT to "0U",
+		U_INT to "0U",
+		U_LONG to "0UL",
+		FLOAT to "0f",
+		DOUBLE to "0.0"
+	)
+
+
+	@TaskAction
+	fun generate() {
+		vectorTypes()
+	}
+
+	private val allComponents = listOf(1 to "x", 2 to "y", 3 to "z", 4 to "w")
+	private val arithmetics = listOf("plus" to "+", "minus" to "-", "times" to "*", "div" to "/")
+
+	private fun vectorTypes() {
+		for (componentCount in 2..4) {
+			val components = allComponents.take(componentCount)
+
+			buildFile(packageName, "Vector$componentCount") {
+				import("kotlin.math", "sqrt")
+
+				typeList.forEach { (type, zero) ->
+					val vectorType = ClassName(packageName, "${type.simpleName}Vector$componentCount")
+					val mutableVectorType = ClassName(packageName, "Mutable${vectorType.simpleName}")
+
+					buildClass(vectorType) {
+						modifiers(SEALED)
+
+						components.forEach { (_, name) ->
+							property(name, type, ABSTRACT)
+						}
+
+						function("get") {
+							modifiers(OPERATOR)
+							parameter("index", INT)
+							controlFlow("when (index)") {
+								components.forEach { (i, name) ->
+									statement("${i - 1} -> $name")
+								}
+								statement("else -> throw IndexOutOfBoundsException()")
+							}
+						}
+
+						property("length", type) {
+							getter {
+								val code = when (type) {
+									BYTE, SHORT, INT, LONG ->
+										"return sqrt((this dot this).toDouble()).to${type.simpleName}()"
+									U_BYTE, U_SHORT, U_INT, U_LONG ->
+										"return sqrt((this dot this).toDouble()).to${type.simpleName.drop(1)}().to${type.simpleName}()"
+									else ->
+										"return sqrt(this dot this)"
+								}
+								statement(code)
+							}
+						}
+
+						property("lengthSquared", type) {
+							getter {
+								statement("return this dot this")
+							}
+						}
+
+						function("normalized") {
+							returns(vectorType)
+							statement("val length = length")
+							val args = components.joinToString { (_, name) ->
+								when (type) {
+									BYTE, SHORT,
+									U_BYTE, U_SHORT -> "($name / length).to${type.simpleName}()"
+									else -> "$name / length"
+								}
+							}
+							statement("return %T($args)", vectorType)
+						}
+
+						function("dot") {
+							modifiers(INFIX)
+							parameter("other", vectorType)
+							returns(type)
+							val result = components.joinToString(" + ") { (_, name) -> "$name * other.$name" }
+								.let {
+									when (type) {
+										BYTE, SHORT,
+										U_BYTE, U_SHORT -> "($it).to${type.simpleName}()"
+										else -> it
+									}
+								}
+							statement("return $result")
+						}
+
+						if (componentCount == 3) {
+							function("cross") {
```

Comment:
This one too should be top-level.

Reply:
https://github.com/Dominaezzz/kgl/pull/19/commits/f7b4dfa6f636fd8074adcaaa8e246e4f5d5e12a4